File: leetcode/BinaryTreeInorderTraversal.cpp
//
// http://www.leetcode.com/onlinejudge#
// 3sum.
// @Author: Xin Chen
// @Created on: 12/13/2012
// @Last modified: 12/13/2012
//
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
#include <stack>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> inorderTraversal(TreeNode *root) {
        vector<int> v;
        stack<TreeNode *> s;
        if (root == NULL) return v;
    
        TreeNode * n = root;
        while(1) {
            while (n != NULL) {
              s.push(n);
              n = n->left;
            }
    
            if (s.size() > 0) {
                n = s.top();
                s.pop();
                v.push_back(n->val);
                n = n->right;
            } else {
                break;
            }
        }
        return v;
    }
    
    /*       
    // recursive version. This works.
    vector<int> inorderTraversal(TreeNode *root) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        vector<int> v;
        inorder(root, v);
        return v;
    }
    
    void inorder(TreeNode * n, vector<int> &v) {
        if (n == NULL) return;
        inorder(n->left, v);
        v.push_back(n->val);
        inorder(n->right, v);
    }
    */
};

int main() { return 0; }

File: leetcode_java/BinaryTreeInorderTraversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> list = new ArrayList<Integer>();
        Stack<TreeNode> s = new Stack<TreeNode>();
        
        TreeNode n = root;
        while (true) {
            while (n != null) {
                s.push(n);
                n = n.left;
            }
            
            if (s.isEmpty()) {
                break;
            }
            else {
                n = s.peek();
                list.add(n.val);
                s.pop();
                n = n.right;
            }
        }
        
        return list;
    }
}


/**
Binary Tree Inorder Traversal
Difficulty: Medium

Given a binary tree, return the inorder traversal of its nodes' values.

For example:
Given binary tree {1,#,2,3},

   1
    \
     2
    /
   3

return [1,3,2].

Note: Recursive solution is trivial, could you do it iteratively?

confused what "{1,#,2,3}" means? > read more on how binary tree is serialized on OJ.

OJ's Binary Tree Serialization:

The serialization of a binary tree follows a level order traversal, 
where '#' signifies a path terminator where no node exists below.

Here's an example:

   1
  / \
 2   3
    /
   4
    \
     5

The above binary tree is serialized as "{1,2,3,#,#,4,#,#,5}". 
 */
